src/audio/AudioPlayer.ts

```diff
@@ -232,13 +248,48 @@ export class AudioPlayer extends EventEmitter {
 			throw new Error(`Cannot play a resource (${resource.name ?? 'unnamed'}) that has already ended.`);
 		}
 
-		this.state = {
-			status: AudioPlayerStatus.Playing,
-			missedFrames: 0,
-			resource,
-			nextTime: Date.now(),
-			stepTimeout: setTimeout(() => this._step(), 0)
-		};
+		if (resource.playStream.readable) {
+			this.state = {
+				status: AudioPlayerStatus.Playing,
+				missedFrames: 0,
+				resource,
+				nextTime: Date.now()
+			};
+			setImmediate(() => this._step());
+		} else {
+			const onReadableCallback = () => {
+				if (this.state.status === AudioPlayerStatus.Buffering && this.state.resource === resource) {
+					this.state = {
+						status: AudioPlayerStatus.Playing,
+						missedFrames: 0,
+						resource,
+						nextTime: Date.now()
+					};
+					setImmediate(() => this._step());
+				}
+			};
+
+			const onFailureCallback = () => {
+				if (this.state.status === AudioPlayerStatus.Buffering && this.state.resource === resource) {
+					this.state = {
+						status: AudioPlayerStatus.Idle
+					};
+				}
+			};
+
+			resource.playStream.once('readable', onReadableCallback);
+
+			resource.playStream.once('end', onFailureCallback);
+			resource.playStream.once('close', onFailureCallback);
+			resource.playStream.once('finish', onFailureCallback);
```

Comment:
The callbacks could do a clean-up, making `onReadableCallback` unregister the listeners for `end`, `close`, and `finish`. Similarly, `onFailureCallback` could unregister all three (since we don't know which one) and the `readable`.

Reply:
Actually, do the following lines cover this? 

https://github.com/discordjs/voice/blob/eb63ebca6938dd8438759657ed3885e4beb17696/src/audio/AudioPlayer.ts#L198-L204

Or can there be an edge case in which this could leak memory if an event is never emitted and this path is called all the time? e.g. when calling `play` with a stream that's not yet readable multiple times, triggering subsequent buffering states.

Reply:
Yep, the lines you linked will cover unregistering for all the bound events.

I think that a stream with no events would be very unlikely. And if there is such a case, the user will probably eventually call `.play()` with another resource, which will unbind the event listeners.